File: 2019/1201/StackAndQueue/queue.java
public class queue{
    
    protected int[] arr;
    protected int size=0;
    protected int front=0;
    protected int back=0;

    public queue(){
        reassign(10);
    }

    public queue(int size){
        reassign(size);
    }

    public void reassign(int size){
        this.arr=new int[size];
        this.size=0;
        this.front=0;
        this.back=0;
    }

    public int size(){
        return this.size();
    }

    public boolean isEmpty(){
        return this.size==0;
    }

    public void display(){
        for(int i=0;i<this.size;i++){
            int idx=(i+this.front)%this.arr.length;
            System.out.print(arr[idx]+ " ");
        }

        System.out.println();
    }

    protected void push_(int val){
        this.arr[this.back]=val;
        this.back=(this.back+1)%this.arr.length;
        this.size++;
    
    }

    public void push(int val) throws Exception{
        if(this.size==this.arr.length){
            throw new Exception("QueueSizeFull");
        }
         push_(val);
     }

    public int pop() throws Exception{
        if(this.size==0){
            throw new Exception("QueueEmpty");
        }

        int rv=this.arr[this.front];
        this.arr[this.front]=0;
        this.front=(this.front+1)%this.arr.length;
        this.size--;

        return rv;
    }

    public int front() throws Exception{
        if(this.size==0){
            throw new Exception("QueueEmpty");
        }

        return this.arr[this.front];
    }
    
}

File: 2019/levelUp_juneBatch/lecture_007_HashMapAndHeap/Stack/queue.cpp
#include <iostream>

using namespace std;

class queue
{
    int *arr;
    int f;
    int b;
    int Elements;
    int TotalSize;

protected:
    void assignSize(int n)
    {
        this->arr = new int[n];
        this->f = 0;
        this->b = 0;
        this->Elements = 0;
        this->TotalSize = n;
    }

    int capacity()
    {
        return this->TotalSize;
    }

public:
    queue()
    {
        assignSize(5);
    }

    queue(int n)
    {
        assignSize(n);
    }

    bool isEmpty()
    {
        return this->Elements == 0;
    }

    int size()
    {
        return this->Elements;
    }

    void display()
    {
        string str = "[";
        for (int i = 0; i < this->Elements; i++)
        {
            int idx = (this->f + i) % this->TotalSize;
            str += to_string(this->arr[idx]);
            if (i != this->Elements - 1)
                str += ",";
        }
        str += "]";
        cout << str << endl;
    }

    void push(int val)
    {
        if (this->Elements == this->TotalSize)
            return;

        this->arr[this->b] = val;
        this->b = (this->b + 1) % this->TotalSize;
        this->Elements++;
    }

    int front()
    {
        if (this->Elements == 0)
            return -1;

        return this->arr[this->f];
    }

    int pop()
    {
        if (this->Elements == 0)
            return -1;
        int rv = this->arr[this->f];
        this->arr[this->f] = 0;

        this->f = (this->f + 1) % this->TotalSize;
        this->Elements--;

        return rv;
    }
};

void solve()
{
    queue que;
    que.push(10);
    que.push(20);
    que.push(30);
    que.push(40);
    que.push(50);

    que.pop();
    que.pop();
    que.push(60);
    que.push(70);

    que.display();
}

int main()
{
    solve();
    return 0;
}